Code/Tools/AssetProcessor/native/utilities/ApplicationManagerBase.cpp

```diff
@@ -449,50 +449,54 @@ void ApplicationManagerBase::InitFileMonitor()
     {
         const auto cachePath = QDir::toNativeSeparators(cacheRoot.absolutePath());
 
+        // For the handlers below, we need to make sure to use invokeMethod on any QObjects so Qt can queue
+        // the callback to run on the QObject's thread if needed.  The APM methods for example are not thread-safe.
+
         const auto OnFileAdded = [this, cachePath](QString path)
         {
             const bool isCacheRoot = path.startsWith(cachePath);
-            if (isCacheRoot)
-            {
-                m_fileStateCache->AddFile(path);
-            }
-            else
+            if (!isCacheRoot)
             {
-                m_assetProcessorManager->AssessAddedFile(path);
-                m_fileStateCache->AddFile(path);
+                bool result = QMetaObject::invokeMethod(m_assetProcessorManager, "AssessAddedFile", Q_ARG(QString, path));
+                AZ_Assert(result, "Failed to invoke m_assetProcessorManager::AssessAddedFile");
```

Comment:
Is this really a coder's fault here or could the IO device just be throwing an error?
In other words, I just want to make sure that this needs an AZ_Assert.

Reply:
Yeah, so this is only going to return false if the method in the string doesn't exist or if the parameters passed in are the wrong type/number, so definitely a programmer-only error.

Reply:
From a non-coder's point of view, and AZ_Assert() looks like a crash